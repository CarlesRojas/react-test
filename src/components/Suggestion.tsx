import { Suggestion as SuggestionType } from "../contexts/MapsAPI";

interface Props {
    suggestion: SuggestionType;
    i: number;
    selected: boolean;
    onMouseEnter: (i: number) => any;
    onSuggestionClick: (suggestion: SuggestionType) => any;
    query: string;
}

export default function Suggestion({ suggestion, i, selected, onMouseEnter, onSuggestionClick, query }: Props) {
    // Normalize string
    const normalizeString = (value: string) => {
        return value
            .toLowerCase()
            .normalize("NFD")
            .replace(/\p{Diacritic}/gu, "");
    };

    // Get index where the query is found in the description
    const queryIndex = normalizeString(suggestion.name).indexOf(normalizeString(query));

    // #################################################
    //   RENDER
    // #################################################

    return (
        <p className={selected ? "selected" : ""} onMouseEnter={() => onMouseEnter(i)} onClick={() => onSuggestionClick(suggestion)}>
            {suggestion.name.slice(0, queryIndex)}
            <span>{suggestion.name.slice(queryIndex, queryIndex + query.length)}</span>
            {suggestion.name.slice(queryIndex + query.length, suggestion.name.length)}
        </p>
    );
}
