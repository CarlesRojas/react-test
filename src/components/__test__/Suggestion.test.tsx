import React from "react";
import renderer from "react-test-renderer";
import Suggestion from "../Suggestion";

test("Suggestion should render the name with the query part highlighted", () => {
    // Mock Suggestion
    const mockSuggestion = {
        name: "Barcelona",
        lat: 41.386950887735196,
        lng: 2.170084327636621,
        id: "randomID",
    };

    // Create snapshot
    const suggestionJSON = renderer.create(<Suggestion suggestion={mockSuggestion} i={0} selected={true} onMouseEnter={() => null} onSuggestionClick={() => null} query="arcelon" />).toJSON();

    // Compare with previous snapshot
    expect(suggestionJSON).toMatchSnapshot();
});
