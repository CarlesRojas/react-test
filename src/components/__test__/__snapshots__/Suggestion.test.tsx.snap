// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestion should render the name with the query part highlighted 1`] = `
<p
  className="selected"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  B
  <span>
    arcelon
  </span>
  a
</p>
`;
